heron/schedulers/src/java/com/twitter/heron/scheduler/mesos/framework/MesosFramework.java

```diff
@@ -0,0 +1,547 @@
+import java.util.Queue;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.LinkedTransferQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Logger;
+
+import org.apache.mesos.Protos;
+import org.apache.mesos.Scheduler;
+import org.apache.mesos.SchedulerDriver;
+
+import com.twitter.heron.spi.common.Config;
+import com.twitter.heron.spi.common.Context;
+
+/**
+ * Mesos Framework, communicating with Mesos Master, to do the actual scheduling.
+ */
+
+public class MesosFramework implements Scheduler {
+  private static final Logger LOG = Logger.getLogger(MesosFramework.class.getName());
+
+  private final Config heronConfig;
+  private final Config heronRuntime;
+
+  // ContainerIndex -> BaseContainer
+  private final Map<Integer, BaseContainer> containersInfo;
+
+  // ContainerIndex -> TaskId
+  private final Map<Integer, String> tasksId;
+
+  // Thread-safe queue of taskId, representing BaseContainer to launch
+  private final Queue<String> toScheduleTasks;
+
+  // Flag indicates whether terminated
+  private volatile boolean isTerminated;
+
+  // The Mesos' SchedulerDriver
+  private volatile SchedulerDriver driver;
+
+  // Indicates whether the MesosFramework has registered.
+  // Scheduler.registered(...) method will count down and release the latch
+  private final CountDownLatch registeredLatch;
+
+  private volatile Protos.FrameworkID frameworkId;
+
+  public MesosFramework(Config heronConfig, Config heronRuntime) {
+    this.heronConfig = heronConfig;
+    this.heronRuntime = heronRuntime;
+    this.containersInfo = new ConcurrentHashMap<>();
+    this.tasksId = new ConcurrentHashMap<>();
+    this.toScheduleTasks = new LinkedTransferQueue<>();
+
+    // Initialize it as 1. Will count down and release once got registered
+    this.registeredLatch = new CountDownLatch(1);
+    this.isTerminated = false;
+  }
+
+  // Create a topology
+  public boolean createJob(Map<Integer, BaseContainer> jobDefinition) {
+    if (isTerminated) {
+      LOG.severe("Job has been killed");
+      return false;
+    }
+
+    // Record the jobDefinition
+    containersInfo.putAll(jobDefinition);
+
+    // To scheduler them with 0 attempt
+    for (Map.Entry<Integer, BaseContainer> entry : jobDefinition.entrySet()) {
+      Integer containerIndex = entry.getKey();
+      BaseContainer container = entry.getValue();
+      String taskId = TaskUtils.getTaskId(container.name, 0);
+      scheduleNewTask(taskId);
+    }
+
+    return true;
+  }
+
+  // Kill a topology
+  public boolean killJob() {
+    if (isTerminated) {
+      LOG.info("Job has been killed");
+      return false;
+    }
+    isTerminated = true;
+
+    LOG.info(String.format("Kill job: %s", Context.topologyName(heronConfig)));
+    LOG.info("Remove all tasks to schedule");
+    toScheduleTasks.clear();
+
+    // Kill all the tasks
+    for (String taskId : tasksId.values()) {
```

Comment:
Good catch. Added synchronized block: 
```
synchronized (toScheduleTasks) {}
```
in critical blocks inside some public methods to avoid concurrent issues.